### DISCUSSION TITLE
megaTinyCore + wifi101: ATtiny3216SN

### DISCUSSION THREAD

**Discussion by Hai-Elec** [SpenceKonde/megaTinyCore#875] [Library discussion]
https://forum.arduino.cc/t/megatinycore-wifi101-attiny3216sn/1062893

I only need control some buttons and log some data byte to server, so I use ATtiny3216SN with module ATWINC1500, I think ATtiny3216SN 20 pin is very cheap and suitable for all small MCU PCB, (I have built a single touch button PCB with ATtiny3216SN, it worked OK with me).
At first, I use MPLAB X IDE to generate library to connect ATtiny3216SN and ATWIC1500 module, then, after that, I do not know how to do next steps to continue coding for my applicant. I do not understand how to use the sample code of microchip with atmega4808 or SAMD21 for my ATtiny3216.

Then I switch to Arduino DUE R3, I connected DUE to ATWINC1500 module, and it work with sample WebServerWiFi of Arduino WIFI101 library, and DUE can also log data to XAMPP server with coding line from guidance DOC.
Very happy with result on DUE, I continue to use ATtiny3216 and ATWINC1500, (although ATtiny3216 can work with 3.3V but I use it for 5V, so I use shift logic level to connect ATtiny3216 with ATWINC1500 module).
But I have trouble with client.read() command and example WebServerWiFi do not work with ATtiny3216.
I post it to Arduino forum already. Now, I hope someone here, who more familiar with megaTinyCore, can help me.
if you have time, you can check my post on Arduino forum at link below:
https://forum.arduino.cc/t/megatinycore-wifi101-attiny3216sn/1062893

**Comment by SpenceKonde** [comment]
I have not worked with wifi stuff yet (and would be using ESP8266's when I
do, cause they're cheap and I'm cheap.

There;s a port of Ethernet. but I don't think there is of wifi

On Wed, Dec 21, 2022 at 3:32 AM Hai-Elec ***@***.***> wrote:

> https://forum.arduino.cc/t/megatinycore-wifi101-attiny3216sn/1062893
>
> I only need control some buttons and log some data byte to server, so I
> use ATtiny3216SN with module ATWINC1500, I think ATtiny3216SN 20 pin is
> very cheap and suitable for all small MCU PCB, (I have built a single touch
> button PCB with ATtiny3216SN, it worked OK with me).
> At first, I use MPLAB X IDE to generate library to connect ATtiny3216SN
> and ATWIC1500 module, then, after that, I do not know how to do next steps
> to continue coding for my applicant. I do not understand how to use the
> sample code of microchip with atmega4808 or SAMD21 for my ATtiny3216.
>
> Then I switch to Arduino DUE R3, I connected DUE to ATWINC1500 module, and
> it work with sample WebServerWiFi of Arduino WIFI101 library, and DUE can
> also log data to XAMPP server with coding line from guidance DOC.
> Very happy with result on DUE, I continue to use ATtiny3216 and
> ATWINC1500, (although ATtiny3216 can work with 3.3V but I use it for 5V, so
> I use shift logic level to connect ATtiny3216 with ATWINC1500 module).
> But I have trouble with client.read() command and example WebServerWiFi do
> not work with ATtiny3216.
> I post it to Arduino forum already. Now, I hope someone here, who more
> familiar with megaTinyCore, can help me.
> if you have time, you can check my post on Arduino forum at link below:
> https://forum.arduino.cc/t/megatinycore-wifi101-attiny3216sn/1062893
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/875>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW2FHG5CTEGH5DJS53LWOK6A3ANCNFSM6AAAAAATFLPTFA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by Hai-Elec** [reply]
Honestly, I choose expensive ATWINC1500 10USD for my first work WIFI connection project, because I think it is quite expensive enough and it also new product in some years recently.
So, maybe the supplier will make some function for it to deal with very weak signal strength, and maybe they also increase speed of transfer.
Although, I only send some byte and control some buttons, but, I really hope that speed of transfer can be as fast as possible and can work anywhere in workshop, in very weak signal strength. I use ATWINC1500 with external antenna version.
I surf net for some cheap 3$ module ESP8266 and they almost is UART interface, that make me think they maybe slower than ATWINC1500 with SPI interface. I also read some complaint with the WIFI range of ESP8266. I think, maybe because ESP8266 is low cost, so, it will limit hardware of the signal 's range function.

**Reply by SpenceKonde** [reply]
The question of UART vs SPI is an old one. SPI is strictly speaking faster, but at the amounts of data where that matters, you're getting past what you can handle on small AVR. UART is the only one of the big-three serial interfaces that provides a transfer model more consistent with how internet connected sensors often need to communicate: as commands come in from the internet (meaning you want the wifi-connected device to be able to initiate a transfer of data, but the sensor may also detect something that needs to be communicated to the web-side of the system. SPI (which is fast, but pin-hungry) and I2C (which is very slow, but pin efficient) are both master-slave solutions, where one side always has to be the one initiates the transactions, while UART allows either side to talk to the other.

Again, that's why I like being able to have each part of the system doing one part of the job, the part it's good at. 

You might want to consider a larger chip - if you want to stick with the nice modern AVRs and their peripherals that aren't 2 decades old in design, you could use an AVR-Dx-series, from the 14-pin AVR64DD14 - AVR64DD32, only slightly above the pricing of the tinyAVRs, or you could go with the next price point and get an AVR128DB for 128k flash, which comes with 16k of ram. 

The Due of course is also based around a fancier chip by far, and has a hell of a lot more computational grunt than an AVR. 

You'll find less info available for the WINC1500 than the ubiquitous and dirt cheap ESP8266, simply because people a skinflints and buy things that are dirt cheap whenever they can. 
The other advantage of ESP8266's (or the more powerful ESP32) is that you can have each one doing it's own "part" of the job. I made a few devices pairing an AVR with an ESP8266, and programmed both of them, with the ESP8266 getting data from the AVR and telling the avr of any commands, but otherwise handling the webserver (and both of them toggled a pin every second, and could fire the others' reset pin, Meanwhile the AVR (either a DB/DD with MVIO at 3.3V, or a tiny running at 3.3V (the Dx-series is nice in that they don't have voltage dependent maximum speed. They have internal regulators, so even running at 5v, that's just the I/O ports - the CPU core runs at like 1.something V no matter what, and since they can count on this, they can use a core optimized for that specific low voltage and squeeze out more performance from it than they could if  the core had to work over a range of voltages (that's why there's the maximum voltage slew rate between 1.8 and 5.5v - if it ramps insanely fast (like, board with practically no board level decoupling being abruptly connected to power supply with enough capacitance on the output that, compared to the board that it can be assumed to have a constant output voltage when the caps of the board are connected. For example, a PCB with a USB plug made from the PCB, with just the minimum decoupling capacitors getting plugged into a typical laptop computer (hence, short low inductance low resistance path from the heavily decoupled supply to the underdecoupled pcb) can exceed that limit, causing the internal regulator to overshoot and potentially damage the chip; use of board level decoupling, which I consider to be elementary engineering practice, unlike apparently a disturbing number of people) - but that architecture ensures maximum speed is possible at all voltages, and nearly eliminates dependance of speed on voltage (testing the watch-crystal oscillator autotune with just voltage swings proved impossible, I had to use temperature to make it change enough for autotune to nudge the speed (meaning freeze spray or (from a distance, not up close, and not immediately after using flamable freeze spray) a torch, which was able to make the internal oscillator speed change enough that autotune could notice and correct it). Wifi range is hard to quantify _a priori_ and being environmentally dependent, needs to be determined experimentally on site typically, as things like furniture and positioning of the device and router make a difference.

**Comment by Hai-Elec** [comment]
Thanks SpenceKonde for your comments! 
After long Tet holiday in Viet Nam, I back to work.
I also have started with ESP-07 module, because it is a module with external antenna. Later, maybe I will go to esp 32 with Bluetooth function. ESP-07 is used as independent module to be a web server, received data from ATtiny3216. 

I think I should add some comment below, just for someone else need similar project for ATtiny3216:
After I check MPLAB X IDE for ATWINC1500 and ATtiny3216, also take a look for SAMD21-WiFi module , and I understand that I need many time to learn more with code language to programing WiFi module better, need learn more coding skill to connect those ICs together.
Before, I used ATtiny3216 with ATWINC1500, the Arduino code of client log data to sever, those Arduino code take amount 70% flash size of ATtiny3216. And, (if Arduino code for tinycore can work when WiFi web server control button) both log data to server and WiFi web server control button need 90% flash memory area of ATtiny3216.
So, I still have 10% flash size available, it still enough to control speed of 1 motor with PWM, and check sensor data via USART, then send data to WiFi server via SPI port of ATtiny3216, connected to ATWINC1500. That is good fit for my project(all 100% flash size is used :-). AVR DB/DD is good MCU, but my very small project is focus to WiFi function than a host MCU. 
I prefer ATtiny3216 because I used and I think it 's  touch capacitive sensor/point function is so much better than AVR MCU(Atmega328PB). Recently, most of my project is touch capacitive sensor/point with Microchip Studio.
So, I think if I can use multiple ATtiny3216 to work together, I will not need to change MCU if project requirement change, I going to use only 1 type MCU ATtiny3216, three or four ATtiny3216 support together, until the level of project need change to SAM3X8E MCU of Arduino DUE board. I mean I want to make a group of ATtiny3216 for small and middle size project. I very rarely to work with big project, until now, I only have 1 simple big project, use almost all port IO of SAM3X8E with Arduino IDE coding.
